File: salary.r
# 4
library(readr)
# Importing the dataset
sd.at <- read.csv('C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\Salary_Data.csv', header = T)
View(sd.at)
# Performing EDA
summary(sd.at)
library(Hmisc)
describe(sd.at)
library("lattice") 
# Graphical exploration
# Dotplot
dotplot(sd.at$YearsExperience, main = "Dot Plot of Waist Circumferences")
dotplot(sd.at$Salary, main = "Dot Plot of Adipose Tissue Areas")
# Boxplot
boxplot(sd.at$YearsExperience, col = "dodgerblue4")
boxplot(sd.at$Salary, col = "red", horizontal = T)
# Histogram
hist(sd.at$YearsExperience)
hist(sd.at$Salary)
# Normal QQ plot
qqnorm(sd.at$YearsExperience)
qqline(sd.at$YearsExperience)
qqnorm(sd.at$Salary)
qqline(sd.at$Salary)
hist(sd.at$Salary, prob = TRUE)          # prob=TRUE for probabilities not counts
lines(density(sd.at$Salary))             # add a density estimate with defaults
lines(density(sd.at$Salary, adjust = 2), lty = "dotted")   # add another "smoother" density
hist(sd.at$YearsExperience, prob = TRUE)          # prob=TRUE for probabilities not counts
lines(density(sd.at$YearsExperience))             # add a density estimate with defaults
lines(density(sd.at$YearsExperience, adjust = 2), lty = "dotted")   # add another "smoother" density
# Bi-variate analysis
# Scatter plot
plot(sd.at$Salary,sd.at$YearsExperience, main = "Scatter Plot", col = "Dodgerblue4", col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Salary_hike", ylab = "Churn_out_rate", pch = 20)  

# Correlation Coefficient
cor(sd.at$Salary, sd.at$YearsExperience)
# Co-variance
cov(sd.at$Salary, sd.at$YearsExperience)
# Linear Regression model
reg <- lm(sd.at$Salary~ sd.at$YearsExperience, data = sd.at) 
summary(reg)
confint(reg, level = 0.95)
pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)
View(pred)
# ggplot for adding Regression line for data
library(ggplot2)
ggplot(data = sd.at, aes(sd.at$YearsExperience, sd.at$YearsExperience) ) + geom_point() + stat_smooth(method = lm, formula = y ~ x)
# Evaluation the model for fitness 
cor(pred$fit,  sd.at$YearsExperience)
reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse
# Transformation Techniques
# input = log(x); output = y
plot(log(sd.at$Salary), sd.at$YearsExperience)
cor(log(sd.at$Salary), sd.at$YearsExperience)
reg_log <- lm(sd.at$YearsExperience ~ log(sd.at$Salary), data = sd.at)
summary(reg_log)
confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")
pred <- as.data.frame(pred)
cor(pred$fit, sd.at$YearsExperience)
rmse <- sqrt(mean(reg_log$residuals^2))
rmse
# Regression line for data
ggplot(data = sd.at, aes(log(sd.at$Salary), sd.at$YearsExperience) ) + geom_point() + stat_smooth(method = lm, formula = y ~ log(x))
# Log transformation applied on 'y'
# input = x; output = log(y)
plot(sd.at$Salary, log(sd.at$YearsExperience))
cor(sd.at$Salary, log(sd.at$YearsExperience))
reg_log1 <- lm(log(sd.at$YearsExperience) ~ sd.at$Salary, data = sd.at)
summary(reg_log1)
predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))
# Exponential function
pred <- exp(predlog)   
pred <- as.data.frame(pred)
cor(pred$fit, sd.at$YearsExperience)
res_log1 = sd.at$YearsExperience - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse
# Regression line for data
ggplot(data = sd.at, aes( sd.at$Salary, log(sd.at$YearsExperience )) ) + geom_point() + stat_smooth(method = lm, formula = log(y) ~ x)
# Non-linear models = Polynomial models
# input = x & x^2 (2-degree) and output = log(y)
reg2 <- lm(log(sd.at$YearsExperience ) ~ sd.at$Salary + I(sd.at$Salary*sd.at$Salary), data = sd.at)
summary(reg2)
predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)
pred <- as.data.frame(pred)
cor(pred$fit,sd.at$YearsExperience)
res2 = sd.at$YearsExperience - pred$fit
rmse <- sqrt(mean(res2^2))
rmse
# Regression line for data
ggplot(data = sd.at, aes(sd.at$Salary, log(sd.at$YearsExperience)) ) + geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))
# Data Partition
train <- sd.at[1:15, ]
test <- sd.at[16:30, ]
log(train$YearsExperience)
log(test$YearsExperience)
plot(train$Salary, log(train$YearsExperience))
plot(test$Salary, log(test$YearsExperience))
model <- lm(log(sd.at$YearsExperience) ~ sd.at$Salary + I(sd.at$Salary * sd.at$Salary), data = train)
summary(model)
confint(model,level=0.95)
log_res <- predict(model,interval = "confidence", newdata = test)
predict_original <- exp(log_res) # converting log values to original values
predict_original <- as.data.frame(predict_original)
test_error <- test$YearsExperience- predict_original$fit # calculate error/residual
test_error
test_rmse <- sqrt(mean(test_error^2))
test_rmse
log_res_train <- predict(model, interval = "confidence", newdata = train)
predict_original_train <- exp(log_res_train) # converting log values to original values
predict_original_train <- as.data.frame(predict_original_train)
train_error <- train$YearsExperience - predict_original_train$fit # calculate error/residual
train_error
train_rmse <- sqrt(mean(train_error^2))
train_rmse

File: salary.py
# 4
import pandas as pd 
import numpy as np  
# importing the dataset
salary_data = pd.read_csv("C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Linear Regression\\Salary_Data.csv")
# Performing EDA
salary_data.describe()
# Graphical Representation
import matplotlib.pyplot as plt 
# Barplot
plt.bar(height = salary_data.Salary, x = np.arange(1, 31, 1))
# Histogram
plt.hist(salary_data.Salary) 
# Boxplot
plt.boxplot(salary_data.Salary) 
plt.bar(height = salary_data.YearsExperience, x = np.arange(1, 31, 1))
plt.hist(salary_data.YearsExperience) #histogram
plt.boxplot(salary_data.YearsExperience) #boxplot
# Scatter plot
plt.scatter(x = salary_data['YearsExperience'], y = salary_data['Salary'], color = 'green') 
# correlation
np.corrcoef(salary_data.YearsExperience, salary_data.Salary) 
# Covariance
cov_output = np.cov(salary_data.YearsExperience, salary_data.Salary)[0, 1]
cov_output
salary_data.cov()
import statsmodels.formula.api as smf
# Simple Linear Regression
model = smf.ols('Salary ~ YearsExperience', data = salary_data).fit()
model.summary()
pred1 = model.predict(pd.DataFrame(salary_data['YearsExperience']))
# Regression Line
plt.scatter(salary_data.YearsExperience, salary_data.Salary)
plt.plot(salary_data.YearsExperience, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res1 = salary_data.Salary - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1
# Model building on Transformed Data
# Log Transformation
# x = log(YearsExperience); y = Salary
plt.scatter(x = np.log(salary_data['YearsExperience']), y = salary_data['Salary'], color = 'brown')
np.corrcoef(np.log(salary_data.YearsExperience), salary_data.Salary) #correlation
model2 = smf.ols('Salary ~ np.log(YearsExperience)', data = salary_data).fit()
model2.summary()
pred2 = model2.predict(pd.DataFrame(salary_data['YearsExperience']))
# Regression Line
plt.scatter(np.log(salary_data.YearsExperience), salary_data.Salary)
plt.plot(np.log(salary_data.YearsExperience), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res2 = salary_data.Salary - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2
# Exponential transformation
# x = YearsExperience; y = log(Salary)
plt.scatter(x = salary_data['YearsExperience'], y = np.log(salary_data['Salary']), color = 'orange')
np.corrcoef(salary_data.YearsExperience, np.log(salary_data.Salary)) #correlation
model3 = smf.ols('np.log(Salary) ~ YearsExperience', data = salary_data).fit()
model3.summary()
pred3 = model3.predict(pd.DataFrame(salary_data['YearsExperience']))
pred3_at = np.exp(pred3)
pred3_at
# Regression Line
plt.scatter(salary_data.YearsExperience, np.log(salary_data.Salary))
plt.plot(salary_data.YearsExperience, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res3 = salary_data.Salary - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3
# Polynomial transformation
# x = YearsExperience; x^2 = YearsExperience*YearsExperience; y = log(Salary)
model4 = smf.ols('np.log(Salary) ~ YearsExperience + I(YearsExperience*YearsExperience)', data = salary_data).fit()
model4.summary()
pred4 = model4.predict(pd.DataFrame(salary_data))
pred4_at = np.exp(pred4)
pred4_at
# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = salary_data.iloc[:, 0:1].values
X_poly = poly_reg.fit_transform(X)
y = salary_data.iloc[:, 1].values
plt.scatter(salary_data.YearsExperience, np.log(salary_data.Salary))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()
# Error calculation
res4 = salary_data.Salary - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4
# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse
# The best model
from sklearn.model_selection import train_test_split
train, test = train_test_split(salary_data, test_size = 0.3)
finalmodel = smf.ols('np.log(Salary) ~ YearsExperience + I(YearsExperience*YearsExperience)',  data= train).fit()
finalmodel.summary()
# Predict on test Salary
test_pred = finalmodel.predict(pd.DataFrame(test))
pred_test_Salary = np.exp(test_pred)
pred_test_Salary
# Model Evaluation on Test data
test_res = test.Salary - pred_test_Salary
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse
# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
pred_train_AT = np.exp(train_pred)
pred_train_AT
# Model Evaluation on train data
train_res = train.Salary - pred_train_AT
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse

